### PR TITLE
Expose more information about the type of a conversion from IConversionExpression

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes: https://github.com/dotnet/roslyn/issues/10447

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler @JohnHamby

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 1]
```diff
@@ -841,10 +841,36 @@ public interface IConversionExpression : IHasOperatorMethodExpression
...
+
+        /// <summary>
+        /// Returns true if the conversion is an implicit nullable conversion or explicit nullable conversion.
+        /// </summary>
+        bool IsNullable { get; }
+
+        /// <summary>
+        /// Returns true if the conversion is an implicit reference conversion or explicit reference conversion.
+        /// </summary>
+        bool IsReference { get; }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: I find this interface too noisy, to many members to expose information about conversion. Can we include all this information into value returned by ConversionKind property? Perhaps make it a combination of bit flags.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 2]
```diff
@@ -841,10 +841,36 @@ public interface IConversionExpression : IHasOperatorMethodExpression
...
+
+        /// <summary>
+        /// Returns true if the conversion is an identity conversion.
+        /// </summary>
+        bool IsIdentity { get; }
+
+        /// <summary>
+        /// Returns true if the conversion is an implicit ('widening' in VB) conversion.
+        /// </summary>
+        bool IsImplicit { get; }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: It feels that this member can be a source of confusion. It is not an opposite to the already existing property IsExplicit. Both can return true, but the names might suggest otherwise.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Alternatively, we can add one new property that returns a different new interface with all these properties. It will be possible to use it in other places where we might want to expose extended information about a conversion.

File: src/Compilers/VisualBasic/Portable/Semantics/Conversions.vb [hunk 3]
```diff
@@ -23,6 +23,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Private ReadOnly _convKind As ConversionKind
         Private ReadOnly _method As MethodSymbol
 
+        Friend Sub New(kind As ConversionKind)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Semantics/Conversions.vb]: Please remove this constructor or add another required parameter for MethodSymbol. We want developers to think what should be passed to the constructor rather than having a convenient way to drop important information.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 4]
```diff
@@ -1024,6 +1096,42 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Get
         End Property
 
+        Private ReadOnly Property Conversion As Conversion
+            Get
+                Return New Conversion(Me.ConversionKind)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: Should not be dropping method symbol information. See how SemanticModel creates Conversion instance for a BoundConversion node.

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: Should not be dropping method symbol information. See how SemanticModel creates Conversion instance for a BoundConversion node.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 5]
```diff
@@ -932,6 +932,42 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+            End Get
+        End Property
+
+        Private ReadOnly Property IConversionExpression_IsIdentity As Boolean Implements Semantics.IConversionExpression.IsIdentity
+            Get
+                Return Me.Conversion.IsIdentity
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: Conversion structure is not used by VB compiler itself, it is created only as an output for external consumers. We should keep it this way. All the answers could be answered directly off of Me.ConversionKind, see how Conversion structure does this, for example. There is no need to add an additional overhead.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 6]
```diff
@@ -1070,6 +1178,36 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
 
+        Private ReadOnly Property IConversionExpression_IsIdentity As Boolean Implements Semantics.IConversionExpression.IsIdentity
+            Get
+                Return False
+            End Get
+        End Property
+
+        Private ReadOnly Property IConversionExpression_IsImplicit As Boolean Implements Semantics.IConversionExpression.IsImplicit
+            Get
+                Return False
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: Why is this the right value to return.

**Comment by AlekseyTs** [comment]
There are no tests. Every new code path in the compilers' codebase should be covered by a unit-test.

**Comment by AlekseyTs** [comment]
Should this change target master branch?

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: > Perhaps make it a combination of bit flags.

I would be very wary of doing this through bitflags.  That's not how we've traditionally exposed any compiler or IDE information.  BitFlags seem great as an internal storage decision.  Not as something people have to comprehend on their own.

> Alternatively, we can add one new property that returns a different new interface with all these properties

That's potentially doable.  @JohnHamby What do you think?

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: That's why i asked about " Is there a difference between 'IsExplicit' and 'ExplicitCastInCode'?"

It sounds like there is.  If so, i recommend renaming IsExplicit to something more like "IsCast" (or "HasSourceCast"), and then having an actual IsExplicit that maps to IsExplicit (or 'IsNarrowing').

File: src/Compilers/VisualBasic/Portable/Semantics/Conversions.vb (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Semantics/Conversions.vb]: Sure.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: Sure.

**Comment by JohnHamby** [comment]
I applaud @CyrusNajmabadi for diving into this, and request another day to review it. I'm out sick and need to give this more consideration than my head cold permits.

**Comment by CyrusNajmabadi** [comment]
> There are no tests

Yup.  I wanted to wait on tests until there was more approval about if this was hte right way to go codewise.  Didn't want to write a bunch of a tests that i would hten have to fixup if i had to majorly refactor this :)

**Comment by CyrusNajmabadi** [comment]
> and request another day to review it.

Absolutely.  There isn't any rush on this.

**Comment by CyrusNajmabadi** [comment]
Ok, i've taken this in another direction that i think is a bit cleaner (both in terms of hte API and in terms of the impl).

I'd like @JohnHamby and @mattwar to give their thoughts on this and on the feedback @AlekseyTs has made.  Aleksey has given some very good feedback about this area and thoughts about using things like Enums+Extension methods instead.  I'd like others to weigh in on what they think the right shape that this portion of the IOperation should have.

Thanks!

**Comment by CyrusNajmabadi** [comment]
> Should this change target master branch?

Probably not :)

This shoudl go into whatever Dev15 will be.  I've put it here now since the branches seem like they're in flux.  Once we're ready for this to go in, i'll get it in the right branch.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 2 above)
Comment (mattwar) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: C# distinguishes between implicit an explicit user defined conversions. But being one or the other does not necessarily mean it was represented in source. It is easy to be confusing if you start using the term explicit to mean written in the source.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: I agree.  I think IConversionExpression.IsExplicit shoudl be renamed.  SOmething like "HasSourceCast" or "HasCastInSource" might be more appropriate.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 7]
```diff
@@ -828,6 +828,54 @@ public static BinaryOperandsKind GetBinaryOperandsKind(BinaryOperationKind kind)
...
+
+        /// <summary>
+        /// Returns true if the conversion is an implicit user-defined conversion or explicit user-defined conversion.
+        /// </summary>
+        bool IsUserDefined { get; }
+
+        /// <summary>
+        /// Returns true if the conversion is an implicit default ('null' in C#, 'nothing' in VB) literal conversion.
+        /// </summary>
+        bool IsDefault { get; }
```
Comment (mattwar) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Why is 'default' associated with null literal conversions?

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: I was trying to pick a good name here.  VB calls it  'IsDefault' and associates it with the 'Nothing' literal.  I was trying ot think of a good name to represent nothing in VB and null in C# and i went with IsDefault since we already call it that elsewhere.

I'm happy to change this to a better name (or we could just remove this for now).   Do you have a recommendation?

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 8]
```diff
@@ -828,6 +829,74 @@ public static BinaryOperandsKind GetBinaryOperandsKind(BinaryOperationKind kind)
...
+
+        /// <summary>
+        /// Returns the method that defines the user defined conversion, if any..
+        /// </summary>
+        IMethodSymbol MethodSymbol { get; }
+    }
+
+    public static class ConversionExtensions
+    {
+        public static IConversion ClassifyConversion(
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Note: i wanted to originally add these to Compilation and SemanticModel.  However, this caused a source level break because we already have extensions with these names defined in the C# and VB compiler dlls.

**Comment by CyrusNajmabadi** [comment]
Tagging @DustinCampbell as well.  This change allows us to share a lot more code in the IDE layer around conversion analysis.  I've done a bit of the work here to demonstrate that, but i wanted dustin to be aware as this sort of common API could likely make a ton of stuff around Reduce/Expand much simpler.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 9]
```diff
@@ -710,7 +710,9 @@ Semantics.ConversionKind IConversionExpression.ConversionKind
             }
         }
 
-        bool IConversionExpression.IsExplicit => this.ExplicitCastInCode;
+        public IConversion Conversion => new Conversion(this.ConversionKind, this.SymbolOpt, isExtensionMethod: false);
+
+        bool IConversionExpression.ExplicitCastInSource => this.ExplicitCastInCode;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Note: i picked a different name for the IConversionExpressoin property because we already use "InSource" in other parts of that API to indicate this concept.  We might want to rename the internal property to match.  If anyone wants me to, i'm happy to do it.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: > Alternatively, we can add one new property that returns a different new interface with all these properties

I have gone ahead and made that change.  i think it does make things much nicer.  It also makes things more uniform in the API wrt to the ```Conversion``` types.  i.e. the existing ```Conversion``` types *are* implementations of the new ```IConversion``` interface.  So things feel very uniform if you're using the semantic model API or if you're using the IOperation API.  The major difference is that IOperation just exposes less info (which is how IOperation normally works).

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 10]
```diff
@@ -828,6 +829,80 @@ public static BinaryOperandsKind GetBinaryOperandsKind(BinaryOperationKind kind)
...
+        /// </summary>
+        bool IsDefault { get; }
+
+        /// <summary>
+        /// Returns the method that defines the user defined conversion, if any..
+        /// </summary>
+        IMethodSymbol MethodSymbol { get; }
+    }
+
+    public static class ConversionExtensions
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: This type definitely doesn't belong in this file. Don't we already have types where we put compilation and semantic model extensions?

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 11]
```diff
@@ -837,14 +912,24 @@ public interface IConversionExpression : IHasOperatorMethodExpression
...
         /// </summary>
         ConversionKind ConversionKind { get; }
+
         /// <summary>
-        /// True if and only if the conversion is indicated explicity by a cast operation in the source code.
+        /// Gets the <see cref="IConversion"/> instance that provides more information about what
+        /// sort of conversion this was.
         /// </summary>
-        bool IsExplicit { get; }
+        IConversion Conversion { get; }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Since this also returns a method symbol, it creates a duplication of information since IConversionExpression also implements IHasOperatorMethodExpression, which is supposed to expose the same information about the method. Perhaps IConversionExpression should no longer inherit from IHasOperatorMethodExpression.

File: src/Compilers/Core/Portable/Compilation/SemanticModel.cs [hunk 12]
```diff
@@ -76,6 +76,10 @@ public IOperation GetOperation(SyntaxNode node, CancellationToken cancellationTo
 
         protected abstract IOperation GetOperationCore(SyntaxNode node, CancellationToken cancellationToken);
 
+        protected internal abstract IConversion GetConversionCore(SyntaxNode expression, CancellationToken cancellationToken);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/SemanticModel.cs]: Doc comment?

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 13]
```diff
@@ -828,6 +829,80 @@ public static BinaryOperandsKind GetBinaryOperandsKind(BinaryOperationKind kind)
         }
     }
 
+    public interface IConversion
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Doc comment?

File: src/Compilers/Core/Portable/Compilation/SemanticModel.cs (hunk 12 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/SemanticModel.cs]: And for other new APIs

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 14]
```diff
@@ -710,7 +710,9 @@ Semantics.ConversionKind IConversionExpression.ConversionKind
             }
         }
 
-        bool IConversionExpression.IsExplicit => this.ExplicitCastInCode;
+        public IConversion Conversion => new Conversion(this.ConversionKind, this.SymbolOpt, isExtensionMethod: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I am not sure this is going to produce correct result for all possible situations we need to handle, at least code in SemanticModel looks more complex than this. We should carefully match the way SemanticModel creates Conversion instances for different scenarios and have unit-tests to test them.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs [hunk 15]
```diff
@@ -14,7 +15,7 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// Summarizes whether a conversion is allowed, and if so, which kind of conversion (and in some cases, the
     /// associated symbol).
     /// </summary>
-    public struct Conversion : IEquatable<Conversion>
+    public struct Conversion : IEquatable<Conversion>, IConversion
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs]: The contract of MethodSymbol doesn't match contract of IConversion.MethodSymbol, it cannot implement the interface.

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs [hunk 16]
```diff
@@ -8,7 +8,7 @@
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
-    internal partial class SymbolDisplayVisitor : AbstractSymbolDisplayVisitor<SemanticModel>
+    internal partial class SymbolDisplayVisitor : AbstractSymbolDisplayVisitor
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs]: Is this change related to introduction of IConversion?

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 17]
```diff
@@ -1018,12 +1030,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
-        Private ReadOnly Property IConversionExpression_IsExplicit As Boolean Implements IConversionExpression.IsExplicit
+        Private ReadOnly Property IConversionExpression_ExplicitCastInSource As Boolean Implements IConversionExpression.ExplicitCastInSource
             Get
                 Return Me.ExplicitCastInCode
             End Get
         End Property
 
+        Private ReadOnly Property IConversionExpression_Conversion As IConversion Implements IConversionExpression.Conversion
+            Get
+                Return New Conversion(Me.ConversionKind, method:=Nothing)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: I believe method cannot be null for user-defined conversions

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 18]
```diff
@@ -1064,7 +1082,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Get
         End Property
 
-        Private ReadOnly Property IConversionExpression_IsExplicit As Boolean Implements IConversionExpression.IsExplicit
+        Private ReadOnly Property IConversionExpression_Conversion As IConversion Implements IConversionExpression.Conversion
+            Get
+                Return New Conversion(ConversionKind.UserDefined, Me.Call.Method)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: BoundUserDefinedConversion is not supposed to implement IConversionExpression, see #8872.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 17 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: See #8872 for more information.

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs (hunk 16 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs]: Please move this change to a separate PR, it looks unrelated to the IOperation/IConversion feature.

**Comment by AlekseyTs** [comment]
I am done with review pass for changes under src/Compilers

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 10 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: > Don't we already have types where we put compilation and semantic model extensions?

Not in this project we don't.   Well, we do have "ModelExtensions", however that lives in Microsoft.CodeAnalysis.  I can't add these extensions to that namespace as it would cause source level breaking changes as they would be seen at the same time as the extnesions we introduce there in the C# and VB compilation dlls.

I've worked around that by putting these into the Microsoft.CodeAnalysis.Semantics namespace.  This has two benefits:

1. No one is pulling this namespace in today, so the introduction of these extensions won't break any source.
2. You get this when you want to work with IConversion.  IConversion is in the "Microsoft.CodeAnalysis.Semantics" namespace, so it makes sense that tha'ts what you would import.

I can pull this into it's own file, but i didn't see a lot of value in it.  Here is where we define IConversion, so it seemed reasonable that hte extensions that would

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 11 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: I would be totally ok with that.  @JohnHamby  are you ok with IConversionExpression not implementing IHasOperatorMethodExpression?

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 17 above)
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: I was basing this on the fact that "IHasOperatorMethodExpression_OperatorMethod" returns nothing below.  This was keepign things consistent with that.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 18 above)
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: I'll let the IOperation owners make that change.  That seems out of scope of what i'm trying to provide here.  I don't jsut want to make the bound node not implement IConversionExpression as clearly more work will need to be done past that.  I'm not familiar enough with the bound nodes and whatnot to do that work as effectively or efficiently as others will be.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 17 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: This is not correct and you will not be able to merge your change with incorrect implementation. Remember, you will have to test the change and make sure implementation is correct.

Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: I don't understand the reasoning here.  It sounds like the implementation is *already* incorrect (as per #8872 ).  It seems like when that issue is fixed we can address the concern you have here with this.  It doesn't make sense to me that we would have to gate fleshing out the API appropriately with also needing to fix other pre-existing issues.

As long as this is properly tracked, it seems like it would be not an issue to get this in.  @srivatsn / @JohnHamby  What do you think?

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: It is very simple all new code (including implementation of new APIs) should be tested and have correct implementation. Spreading bugs around is not an option. The fact that other feature that you happened to base the code off of has zero tests and has incorrect implementation is not an excuse. Also, in order to merge any code into compilers you need a green light from two members of the compiler team. You won't get this sign-off without adequate testing that should involve all possible kinds of conversions as well as making sure that the Conversion instances that you produce are exactly equivalent to those produced by SemanticModel for the same scenarios.

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: See tests in src/Compilers/VisualBasic/Test/Semantic/Diagnostics/IOperationTests.vb as examples of the test we expect to see.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 13 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: I'm not sure that this interface is necessary. Most, if not all, of its properties can be implemented--however awkwardly--as extension methods over IConversionExpression. The ones that cannot be implemented efficiently as extension methods can inspire improvements to IConversionExpression.

I understand that the introduction of the interface is a reaction to earlier review comments, but another iteration on life without it might be worthwhile.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: That was another option that I suggested and I would prefer it over the introduction of the new interface.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 17 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: Perhaps "inadequate testing" is more accurate than  "zero tests", because there are some analyzer tests that exercise IConversionExpression a bit, but I am the one guilty of checking in an inadequately-tested prototype. We are in the process of formalizing a plan for getting adequate testing, and I am not sure how heavy a requirement should be placed on a change like this. My inclination is to give a green light to a checkin that makes things better rather than require that it make things perfect, but the compiler team sets its own policies.

I will help with the testing and bug fixing after we come to agreement on the API changes.

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: @JohnHamby I apologize for exaggeration/generalization. What I meant to say is that there is a lack of testing specifically for user-defined conversions.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 13 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: 1. Can you explain how these extension methods would work?  I can't see how, given the definition of IConversionExpression, how an extension method could answer the questions that i'm exposing through properties.
2. The value of hte interface is that it provides a *common* way to talk about conversions, regardless of if they come through IOperation or if they come through SemanticModel/Compilation.  

The latter is very important to the IDE.  We don't want to handle conversions differently.   It would be like having IOperation return a new entity that represented Symbols.  That would be very awkward and unpleasant.  We would far prefer to avoid that.

**Comment by CyrusNajmabadi** [comment]
I do not intend this to go in without tests.  I have simply avoided tests *until* i am reasonably sure that people are ok with the general implementation approach.  Otherwise i'd end up having to rewrite tons of tests as i adapted the implementation.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 13 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: I cannot speak for @JohnHamby. My suggestion was to incorporate information about language specific conversion kind into the value returned by 
```
ConversionKind ConversionKind { get; }
``` 

And then add convenience extension methods for ConversionKind: IsImplicit, IsIdentity, etc.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: If we do that, then we will still have two different ways of talking about conversions.  We'll have the SemanticModel/Compilation APIs that talk about conversions, and which talk about them differently than how IOperation talks about it.  This means on the IDE side we have to duplicate logic around conversions depending on how we obtained it from the compiler APIs.  This seems very ungainly and works against our goals of common, composable, APIs across VB and C#.  

I'd really like some design feedback on if this is the way we want to go.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: I understand you argument and simply responding to your request to explain my proposal in details.

Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: The argument about having a common treatment of conversions between IOperation and SemanticModel is compelling. I'll go look at the treatment in SemanticModel.

Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: After looking at the C# and VB Conversion type, I understand that this approach is prior art. An IConversion interface that represents a subset of the properties of the preexisting types is a good thing.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: We can either retain the new property or have IConversionExpression derive from IConversion.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 2 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: The terms "implicit" and "explicit" are confusing because they have meaning in one of our source languages with respect to possible conversions, and another meaning with respect to actual conversions.

"PresentInSource" or "HasCastInSource" strike me as OK for indicating that a conversion is explicitly indicated in source, and "IsNarrowing" and "IsWidening" string me as OK for indicating that a conversion is narrowing or widening.

IsNarrowing does not imply PresentInSource for VB.

One could argue that whether or not a cast is present in source is not semantically significant.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: I would be fine with picking either IsExplicit/IsImplicit or IsNarrowing/IsWidening.  Usually, it seems like for the shared codebases, we tend to follow C# naming.  i.e. we have ISwitchStatement not ISelectStatement.  But i'm not picky.

Comment (mattwar) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Unfortunately, the C# naming sort-of implies that the cast was in source or not. Typically implicit conversion operators do not need to be mentioned in source, but explicit ones do. Hence the names. However, for example, being an instance of an implicit conversion does not actually tell you whether it was actually in source or not. So if we use the implicit/explicit terms we tend to confuse ourselves all the time about what we mean.  "Was this cast implicit?"  What question is being asked?

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 7 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: "IsDefaultValue"?

**Comment by CyrusNajmabadi** [comment]
@JohnHamby Do you think you'll be able to take this?  There are definitely IDE features that would benefit from this.

**Comment by JohnHamby** [comment]
Yes, once I am at liberty to work on IOperation.

**Comment by CyrusNajmabadi** [comment]
Ok.  I've merged in the latest stuff.  One thing i'm curious about is this though:

Since IOperation is currently disabled for 3rd parties, could i go ahead and make these conversion changes to unblock IDE scenarios?

**Comment by JohnHamby** [comment]
Making these changes is fine with me. However, the compiler team controls the checkin gate.

**Comment by CyrusNajmabadi** [comment]
Thanks!  I will discuss this with them later this week.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler I'd like to start discussion back up on this PR.  

1. IDE would like to simplify a bunch of code we have that tries to work across VB/C# wrt conversions.  Examples of this include features like "Simplification" as well as potential analyzers that want to advertise themselves based on the conversions they find in code.
2. Original concerns about this PR stemmed from concerns about returning an object to provide information about a conversion (as opposed to something like a packed int/enum + extension methods).  I'd like us to come to consensus on this.  The approach i've taken fits appropriately given the existing APIs that already return conversion structs back.  However, I'm fine with going the int/enum route if we'll also have that same API on the compiler/semantic model side.  That way we have a consistent way of talking about conversions, regardless of if those conversions came from IOperations or Compilation/SemanticModel.
3. There were also concerns about naming and about confusion over terms like "Explicit/Implicit".  Specifically because there's the concept of an 'Explicit' cast, versus a cast that appears explicitly in source.  To help alleviate the confusion the terminology has become "HasExplicitCastInSource" and "IsNarrowing/IsWidening".  While this is VB terminology, it maps nicely to C# and doesn't have the ambiguity over what 'Explicit' means.

I would like to get the teams' thoughts on this.  Especially @mattwar @gafter @AlekseyTs and @JohnHamby .   If we can agree on an appropriate API, then i'm happy to make whatever changes are necessary to update this code to that API.  And once htat is in place, i'm also ok with helping out with the testing here.  (Note that i don't want to do testing preemptively as i don't want to have to fix all of it up if we change the API).  

Also, we would lke to get *something* in sooner rather than later.  Because this API is essentially internal (since you can't actually get at the IOperation stuff through our normal APIs), i feel like we could get something usable in early and then continue to beat on it as we bring IOperation up to fully shippable quality.

Let me know what you all thing.  Also tagging @ManishJayaswal to make him aware of this.

**Comment by CyrusNajmabadi** [comment]
Closing in favor of https://github.com/dotnet/roslyn/pull/18097
